ext/storage_sources/s3_store/s3_store.cpp

```diff
@@ -0,0 +1,253 @@
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <wiredtiger.h>
+#include <wiredtiger_ext.h>
+
+#include <aws/core/Aws.h>
+#include <aws/s3-crt/S3CrtClient.h>
+#include "aws_bucket_conn.h"
+
+#include <iostream>
+
+#define UNUSED(x) (void)(x)
+
+/* S3 storage source structure. */
+typedef struct {
+    WT_STORAGE_SOURCE storage_source; /* Must come first */
+    WT_EXTENSION_API *wt_api;         /* Extension API */
+} S3_STORAGE;
+
+typedef struct {
+    /* Must come first - this is the interface for the file system we are implementing. */
+    WT_FILE_SYSTEM file_system;
+    S3_STORAGE *s3_storage;
+    aws_bucket_conn *conn;
+} S3_FILE_SYSTEM;
+
+typedef struct s3_file_handle {
+    WT_FILE_HANDLE iface; /* Must come first */
+    S3_STORAGE *s3;       /* Enclosing storage source */
+    WT_FILE_HANDLE *fh;   /* File handle */
+} S3_FILE_HANDLE;
+
+/* Configuration variables for connecting to S3CrtClient. */
+Aws::String region = Aws::Region::AP_SOUTHEAST_2;
+const double throughput_target_gbps = 5;
+const uint64_t part_size = 8 * 1024 * 1024; /* 8 MB. */
+
+/* Setting SDK options. */
+Aws::SDKOptions options;
+
+static int s3_customize_file_system(
+  WT_STORAGE_SOURCE *, WT_SESSION *, const char *, const char *, const char *, WT_FILE_SYSTEM **);
+static int s3_add_reference(WT_STORAGE_SOURCE *);
+static int s3_fs_terminate(WT_FILE_SYSTEM *, WT_SESSION *);
+
+/*
+ * s3_customize_file_system --
+ *      Return a customized file system to access the s3 storage source objects.
+ */
+static int
+s3_customize_file_system(WT_STORAGE_SOURCE *storage_source, WT_SESSION *session,
+  const char *bucket_name, const char *auth_token, const char *config,
+  WT_FILE_SYSTEM **file_systemp)
+{
+    S3_FILE_SYSTEM *fs;
+    int ret;
+
+    /* Mark parameters as unused for now, until implemented. */
+    UNUSED(session);
+    UNUSED(bucket_name);
+    UNUSED(auth_token);
+    UNUSED(config);
+
+    Aws::S3Crt::ClientConfiguration aws_config;
+    aws_config.region = region;
+    aws_config.throughputTargetGbps = throughput_target_gbps;
+    aws_config.partSize = part_size;
+
+    if ((fs = (S3_FILE_SYSTEM*)calloc(1, sizeof(S3_FILE_SYSTEM))) == NULL)
+        return (errno);
+
+    fs->s3_storage = (S3_STORAGE*)storage_source;
+
+    /* New can fail; will deal with this later. */
+    fs->conn = new aws_bucket_conn(aws_config);
+    fs->file_system.terminate = s3_fs_terminate;
+
+    /* TODO: Move these into tests. Just testing here temporarily to show all functions work. */
```

Comment:
When you have TODOs, to avoid forgetting them, you can add the comment "FIXME-WT-8698" instead and the script `s_mentions` should pick them up when `s_all` is executed. It basically look for any comment / code that mentions the current ticket number that is retrieved from the branch if I am not mistaken.